### PR TITLE
Added dot product and fir-filter examples.

### PR DESCRIPTION
**Pull request by jbaileyhandle**
Also added int-to-floating-poinnt casting.

### CODE REVIEW

**Comment by jbaileyhandle** [comment]
Bump.

File: xls/dslx/stdlib/apfloat.x [hunk 1]
```diff
@@ -183,6 +183,183 @@ pub fn unflatten<EXP_SZ:u32, SFD_SZ:u32,
   }
 }
 
+pub fn cast_from_fixed<EXP_SZ:u32, SFD_SZ:u32, UEXP_SZ:u32 = EXP_SZ + u32:1,
+  NUM_SRC_BITS:u32, EXTENEDED_SFD_SZ:u32 = SFD_SZ + NUM_SRC_BITS>
```
Comment (RobSpringer) [inline xls/dslx/stdlib/apfloat.x:187]: Spelling.

File: xls/dslx/stdlib/apfloat.x [hunk 2]
```diff
@@ -183,6 +183,183 @@ pub fn unflatten<EXP_SZ:u32, SFD_SZ:u32,
...
+  (to_cast: sN[NUM_SRC_BITS]) -> APFloat<EXP_SZ, SFD_SZ> {
+  // Determine sign.
+  let sign = to_cast[NUM_SRC_BITS-u32:1 : NUM_SRC_BITS];
+
+  // Determine exponent.
+  let abs_magnitude = (to_cast if sign == u1:0 else -to_cast) as uN[NUM_SRC_BITS];
+  let lz = clz(abs_magnitude);
+  let num_trailing_nonzeros = NUM_SRC_BITS as uN[NUM_SRC_BITS] - lz;
+  let exp = num_trailing_nonzeros as uN[UEXP_SZ] - uN[UEXP_SZ]:1;
```
Comment (RobSpringer) [inline xls/dslx/stdlib/apfloat.x:196]: Can you parenthesize this? I (and I assume others?) can never keep precedence rules in my head.

Comment (RobSpringer) [reply xls/dslx/stdlib/apfloat.x:196]: And it looks like you use "uN[UEXP_SZ] - uN[UEXP_SZ]:1" > 1x; consider making a temp to hold it.

File: xls/dslx/stdlib/apfloat.x [hunk 3]
```diff
@@ -183,6 +183,183 @@ pub fn unflatten<EXP_SZ:u32, SFD_SZ:u32,
...
+
+  // Round significand (round to nearest, half to even).
+  let lsb_idx:uN[EXTENEDED_SFD_SZ] = num_trailing_nonzeros as uN[EXTENEDED_SFD_SZ]
+    - uN[EXTENEDED_SFD_SZ]:1;
+  let halfway_idx:uN[EXTENEDED_SFD_SZ] = lsb_idx - uN[EXTENEDED_SFD_SZ]:1;
+  let halfway_bit_mask = uN[EXTENEDED_SFD_SZ]:1 << halfway_idx;
+  let trunc_mask = (uN[EXTENEDED_SFD_SZ]:1 << lsb_idx) - uN[EXTENEDED_SFD_SZ]:1;
+  let trunc_bits = trunc_mask & extended_sfd;
+  let trunc_bits_gt_half = trunc_bits > halfway_bit_mask;
+  let trunc_bits_are_halway = trunc_bits == halfway_bit_mask;
```
Comment (RobSpringer) [inline xls/dslx/stdlib/apfloat.x:215]: Spelling

File: xls/dslx/stdlib/apfloat.x [hunk 4]
```diff
@@ -183,6 +183,183 @@ pub fn unflatten<EXP_SZ:u32, SFD_SZ:u32,
...
+  // Determine significand (pre-rounding).
+  let extended_sfd = abs_magnitude ++ uN[SFD_SZ]:0;
+  let sfd = extended_sfd >>
+    (num_trailing_nonzeros - uN[NUM_SRC_BITS]:1) as uN[EXTENEDED_SFD_SZ];
+  let sfd = sfd[0 : SFD_SZ];
+
+  // Round significand (round to nearest, half to even).
+  let lsb_idx:uN[EXTENEDED_SFD_SZ] = num_trailing_nonzeros as uN[EXTENEDED_SFD_SZ]
+    - uN[EXTENEDED_SFD_SZ]:1;
+  let halfway_idx:uN[EXTENEDED_SFD_SZ] = lsb_idx - uN[EXTENEDED_SFD_SZ]:1;
```
Comment (RobSpringer) [inline xls/dslx/stdlib/apfloat.x:210]: These type annotations shouldn't be needed...but they don't _hurt_ anything either, so...meh.

File: xls/dslx/stdlib/float32.x [hunk 5]
```diff
@@ -35,6 +35,9 @@ pub fn bias(unbiased_exponent_in: u9) -> u8 {
 }
 pub fn flatten(f: F32) -> u32 { apfloat::flatten<u32:8, u32:23>(f) }
 pub fn unflatten(f: u32) -> F32 { apfloat::unflatten<u32:8, u32:23>(f) }
+pub fn cast_from_fixed<NUM_SRC_BITS:u32>(s: sN[NUM_SRC_BITS]) -> F32 {
```
Comment (RobSpringer) [inline xls/dslx/stdlib/float32.x:38]: Any chance you could toss this in the float64 and bfloat16 modules, too?

File: xls/examples/dot_product.x [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import std
```
Comment (RobSpringer) [inline xls/examples/dot_product.x:1]: File comment!

Comment (RobSpringer) [reply xls/examples/dot_product.x:1]: And license header!

File: xls/examples/fir_filter.x [hunk 7]
```diff
@@ -0,0 +1,80 @@
+
```
Comment (RobSpringer) [inline xls/examples/fir_filter.x:1]: File comment & license header!

**Review by RobSpringer — CHANGES_REQUESTED** [review]
Looks really good! Just...gosh, looks like mostly if not all cosmetic stuff.

File: xls/dslx/stdlib/apfloat.x (hunk 1 above)
Comment (jbaileyhandle) [reply xls/dslx/stdlib/apfloat.x:187]: Fixed.

File: xls/dslx/stdlib/apfloat.x (hunk 2 above)
Comment (jbaileyhandle) [reply xls/dslx/stdlib/apfloat.x:196]: Parentheses parenthesized.
For "uN[UEXP_SZ] - uN[UEXP_SZ]:1" , we are casting the value on the left hand side and then subtracting 1.  I don't believe there are common terms here (we cast num_trailing_nonzeros and subtract 1 on line 196 and cast EXP_SZ and subtract 1 on line 197).

File: xls/dslx/stdlib/apfloat.x (hunk 3 above)
Comment (jbaileyhandle) [reply xls/dslx/stdlib/apfloat.x:215]: Fixed (assuming this issue was "EXTENEDED")

File: xls/dslx/stdlib/apfloat.x (hunk 4 above)
Comment (jbaileyhandle) [reply xls/dslx/stdlib/apfloat.x:210]: Removed.

File: xls/dslx/stdlib/float32.x (hunk 5 above)
Comment (jbaileyhandle) [reply xls/dslx/stdlib/float32.x:38]: Done.

File: xls/examples/dot_product.x (hunk 6 above)
Comment (jbaileyhandle) [reply xls/examples/dot_product.x:1]: Done.

File: xls/examples/fir_filter.x (hunk 7 above)
Comment (jbaileyhandle) [reply xls/examples/fir_filter.x:1]: Done.

**Comment by meheff** [comment]
@RobSpringer Is this ready to land?

File: xls/dslx/stdlib/apfloat.x (hunk 3 above)
Comment (RobSpringer) [reply xls/dslx/stdlib/apfloat.x:215]: (it was "halway" -> "halfway") :D

**Review by RobSpringer — APPROVED** [review]
Just one request.

**Comment by meheff** [comment]
I can take care of the spelling correction.

**Pull request merged**